armi/plugins.py

```diff
@@ -120,13 +120,16 @@
    user settings. This also predated the plugin infrastructure, and may one day be
    replaced with plugin-based fuel handler logic.
 """
-from typing import Dict, Union
+from typing import Dict, Union, Callable, Optional, TYPE_CHECKING
 
 import pluggy
 
 from armi import pluginManager
 from armi.utils import flags
 
+if TYPE_CHECKING:
+    from armi.reactor.composites import Composite
+
```

Comment:
Not sure the stance on type hinting but seems positive (#531  / #493). This block is only to help with an IDE showing a type that exists in the function signature, e.g.,
```python
def func(a: "Composite"):
```
does not error if `Composite` is imported in this way. But does if you remove the quotes

Reply:
We are definitely pro-type hinting. For sure.

I'm not thrilled about extra import and `TYPE_CHECKING`.  But perhaps this is the way of the future.  I haven't decided about these details yet.

Still, yes, generally, we like type hinting.

Reply:
Excellent. And by

> I'm not thrilled about extra import

Do you mean the import here of `from armi.reactor.composites import Composite`? If it's a performance issue you're considering, that import will not trigger in a production run (maybe the cause for a coverage drop) because of how python and type checkers like IDEs sets the `TYPE_CHECKING` variable

> A special constant that is assumed to be True by 3rd party static type checkers. It is False at runtime.

So then this import isn't tripped in e.g., `armi run`

https://docs.python.org/3.7/library/typing.html#typing.TYPE_CHECKING

Reply:
Yeah, I was just thinking about performance. 

To be clear, your typing code here is fine. I am just tenuously unhappy with the general idea of having more imports in Python modules, just for the purposes of typing. I think the Python team's design decision seems likely to generally cause more imports and performance hits, to provide typing info. Which... is just insanely backwards.

Again, your code here is fine. I'm just not sure the languages typing solution is a good one.